devnet/devnet.go

```diff
@@ -6,71 +6,77 @@ import (
 	"os"
 	"os/exec"
 	"os/signal"
+	"path/filepath"
 	"strings"
 	"sync"
 	"syscall"
 	"time"
 )
 
-func runShell(ctx context.Context, name string, commands []string) {
+func runCmdsWithLog(ctx context.Context, name string, commands [][]string) {
 	logFile, err := os.Create(name + ".log")
 	if err != nil {
 		log.Fatal(err)
 	}
 	defer logFile.Close()
 
-	src := strings.Join(commands, " && ")
-	cmd := exec.CommandContext(ctx, "sh", "-c", src)
-	cmd.Stdout = logFile
-	cmd.Stderr = logFile
-	if err := cmd.Run(); err != nil {
-		log.Printf("%s; check %s for details", err, logFile.Name())
+	for _, cmdArgs := range commands {
+		log.Printf("command for %s: %s", name, strings.Join(cmdArgs, " "))
+		cmd := exec.CommandContext(ctx, cmdArgs[0], cmdArgs[1:]...)
+		cmd.Stdout = logFile
+		cmd.Stderr = logFile
+		// If ctx.Err()!=nil, we cancelled the command via SIGINT.
+		if err := cmd.Run(); err != nil && ctx.Err() == nil {
+			log.Printf("%s; check %s for derunLotusNodetails", err, logFile.Name())
+			break
+		}
 	}
 }
 
-func runLotusNode(ctx context.Context) {
-	lotusNodeCmds := []string{
-		"lotus-seed genesis new localnet.json",
-		"lotus-seed pre-seal --sector-size 2048 --num-sectors 10",
-		"lotus-seed genesis add-miner localnet.json ~/.genesis-sectors/pre-seal-t01000.json",
-		"lotus daemon --lotus-make-genesis=dev.gen --genesis-template=localnet.json --bootstrap=false",
+func runLotusDaemon(ctx context.Context, home string) {
+	cmds := [][]string{
+		{"lotus-seed", "genesis", "new", "localnet.json"},
+		{"lotus-seed", "pre-seal", "--sector-size=2048", "--num-sectors=10"},
+		{"lotus-seed", "genesis", "add-miner", "localnet.json",
+			filepath.Join(home, ".genesis-sectors", "pre-seal-t01000.json")},
+		{"lotus", "daemon", "--lotus-make-genesis=dev.gen",
+			"--genesis-template=localnet.json", "--bootstrap=false"},
 	}
 
-	runShell(ctx, "lotus-daemon", lotusNodeCmds)
+	runCmdsWithLog(ctx, "lotus-daemon", cmds)
 }
 
-func runMiner(ctx context.Context) {
-	time.Sleep(5 * time.Second) // wait for lotus node to run
-
-	lotusMinerCmds := []string{
-		"lotus wallet import ~/.genesis-sectors/pre-seal-t01000.key",
-		"lotus-miner init --genesis-miner --actor=t01000 --sector-size=2048 --pre-sealed-sectors=~/.genesis-sectors --pre-sealed-metadata=~/.genesis-sectors/pre-seal-t01000.json --nosync",
-		"lotus-miner run --nosync",
+func runLotusMiner(ctx context.Context, home string) {
+	cmds := [][]string{
+		{"lotus", "wait-api"}, // wait for lotus node to run
```

Comment:
note that I also replaced the 5s sleep with this command, following what's done in header.sh.